server/src/main/java/io/spine/server/entity/RepositoryCache.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2019, TeamDev. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.entity;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import io.spine.annotation.Internal;
+import io.spine.server.tenant.IdInTenant;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * The cache of {@code Entity} states for a certain {@code Repository} and
+ * {@linkplain #startCaching(Object) selected} identifiers.
+ *
+ * @param <I>
+ *         the type of {@code Entity} identifiers
+ * @param <E>
+ *         the type of entity
+ */
+@Internal
+public final class RepositoryCache<I, E extends Entity<I, ?>> {
+
+    private final Map<IdInTenant<I>, E> cache = Maps.newConcurrentMap();
+    private final Set<IdInTenant<I>> idsToCache = Sets.newConcurrentHashSet();
+
+    private final boolean multitenant;
+    private final Load<I, E> loadFn;
+    private final Store<I, E> storeFn;
+
+    public RepositoryCache(boolean multitenant, Load<I, E> loadFn, Store<I, E> storeFn) {
+        this.multitenant = multitenant;
+        this.loadFn = loadFn;
+        this.storeFn = storeFn;
+    }
+
+    public E load(I id) {
+        IdInTenant<I> idInTenant = idInTenant(id);
+        if (!idsToCache.contains(idInTenant)) {
+            return loadFn.apply(idInTenant.value());
+        }
+
+        if (!cache.containsKey(idInTenant)) {
+            E entity = loadFn.apply(idInTenant.value());
+            cache.put(idInTenant, entity);
```

Comment:
Can the cache grow too quick and fill the memory?

Reply:
Not in our use cases. The maximum number of cached entities is the total number of shards — and that's in case there is just one node serving the whole app. It's because each shard delivery is a single-thread process, and the only entity is cached at a time by the delivery.  

End-users will hardly ever have 100k of shards and just a single instance processing them all.

I don't want to use Guava's cache for such a simple use case. And it's much slower than the concurrent `HashMap`.

I have added the note on usage though.